File: TwoLayerNetwork.cpp
#include <torch/torch.h>
#include <iostream>

// N is batch size; D_in is input dimension
// H is hidden dimension; D_out is output dimension
const int64_t N = 64;
const int64_t D_in = 1000;
const int64_t H = 100;
const int64_t D_out = 10;

struct TwoLayerNet : torch::nn::Module {
  TwoLayerNet() : linear1(D_in, H), linear2(H, D_out) {
    register_module("linear1", linear1);
    register_module("linear2", linear2);
  }
  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(linear1->forward(x));
    x = linear2->forward(x);
    return x;
  }
  torch::nn::Linear linear1;
  torch::nn::Linear linear2;
};

int main() {
  torch::manual_seed(1);

  torch::Tensor x = torch::rand({N, D_in});
  torch::Tensor y = torch::rand({N, D_out});
  // change this to torch::kCUDA if GPU is available
  torch::Device device(torch::kCPU);

  TwoLayerNet model;
  model.to(device);

  float_t learning_rate = 1e-4;
  torch::optim::SGD optimizer(
      model.parameters(), torch::optim::SGDOptions(learning_rate));

  // number of ecpochs = 500
  for (size_t epoch = 1; epoch <= 500; ++epoch) {
    optimizer.zero_grad();
    auto y_pred = model.forward(x);
    // need to use y.detach() instead of y
    //   see issue: https://github.com/pytorch/pytorch/issues/16830
    auto loss = torch::mse_loss(y_pred, y.detach());
    if (epoch%100 == 99)
      std::cout << "epoch = " << epoch << " " << "loss = " << loss << "\n";
    loss.backward();
    optimizer.step();
  }
}

File: TwoLayerNetwork.py
import torch

class TwoLayerNet(torch.nn.Module):

    def __init__(self, D_in, H, D_out):
        super(TwoLayerNet, self).__init__()
        self.linear1 = torch.nn.Linear(D_in, H)
        self.linear2 = torch.nn.Linear(H, D_out)

    def forward(self, x):
        h_relu = self.linear1(x).clamp(min=0)
        y_pred = self.linear2(h_relu)
        return y_pred

# N is batch size; D_in is input dimension
# H is hidden dimension; D_out is output dimension
N, D_in, H, D_out = 64, 1000, 100, 10

# Create random input and output data
x = torch.randn(N, D_in)
y = torch.randn(N, D_out)

# nn package to define model and loss function
model = TwoLayerNet(D_in, H, D_out)

loss_fn = torch.nn.MSELoss(reduction='sum')
learning_rate = 1e-4
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
for t in range(500):
    # Forward pass: compute predicted y
    y_pred = model(x)

    # Compute and print loss
    loss = loss_fn(y_pred, y)
    if t%100 == 99:
        print(t, loss.item())

    #Zero the gradients before running the backward pass
    optimizer.zero_grad()

    # Backpropagate the loss
    loss.backward()

    # Update weights
    optimizer.step()
